File: app/src/main/java/org/ole/planet/myplanet/model/RealmMyLibrary.java
package org.ole.planet.myplanet.model;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import org.ole.planet.myplanet.MainApplication;
import org.ole.planet.myplanet.ui.sync.SyncActivity;
import org.ole.planet.myplanet.utilities.FileUtils;
import org.ole.planet.myplanet.utilities.JsonUtils;
import org.ole.planet.myplanet.utilities.NetworkUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.RealmResults;
import io.realm.annotations.PrimaryKey;

public class RealmMyLibrary extends RealmObject {
    @PrimaryKey
    private String id;
    private String _id;
    private RealmList<String> userId;
    private String resourceRemoteAddress;
    private String resourceLocalAddress;
    private Boolean resourceOffline = false;
    private String resourceId;
    private String _rev;
    private String downloadedRev;
    private boolean need_optimization;
    private String publisher;
    private String linkToLicense;
    private String addedBy;
    private String uploadDate;
    private long createdDate;
    private String openWith;
    private String articleDate;
    private String kind;
    private String language;
    private String author;
    private String year;
    private String medium;
    private String title;
    private String averageRating;
    private String filename;
    private String mediaType;
    private String resourceType;
    private String description;
    private String sendOnAccept;
    private String translationAudioPath;
    private int sum;
    private int timesRated;
    private RealmList<String> resourceFor;
    private RealmList<String> subject;
    private RealmList<String> level;
    private RealmList<String> tag;
    private RealmList<String> languages;
    private String courseId;
    private String stepId;
    private String downloaded;
    private boolean isPrivate;


    public static List<RealmMyLibrary> getMyLibraryByUserId(Realm mRealm, SharedPreferences settings) {
        RealmResults<RealmMyLibrary> libs = mRealm.where(RealmMyLibrary.class).findAll();
        return getMyLibraryByUserId(settings.getString("userId", "--"), libs, mRealm);
    }


    public static List<RealmMyLibrary> getMyLibraryByUserId(String userId, List<RealmMyLibrary> libs, Realm mRealm) {
        List<RealmMyLibrary> libraries = new ArrayList<>();
        List<String> ids = RealmMyTeam.getResourceIdsByUser(userId, mRealm);

        for (RealmMyLibrary item : libs) {
            if (item.getUserId().contains(userId) || ids.contains(item.getResourceId())) {
                libraries.add(item);
            }
        }
        return libraries;
    }


    public static List<RealmMyLibrary> getMyLibraryByUserId(String userId, List<RealmMyLibrary> libs) {
        List<RealmMyLibrary> libraries = new ArrayList<>();

        for (RealmMyLibrary item : libs) {
            if (item.getUserId().contains(userId)) {
                libraries.add(item);
            }
        }
        return libraries;
    }

    public static List<RealmMyLibrary> getOurLibrary(String userId, List<RealmMyLibrary> libs) {
        List<RealmMyLibrary> libraries = new ArrayList<>();
        for (RealmMyLibrary item : libs) {
            if (!item.getUserId().contains(userId)) {
                libraries.add(item);
            }
        }
        return libraries;
    }

//
//    private static boolean containsId(RealmObject item, Class c,String userId) {
//        if (c == RealmMyCourse.class){
//           return ((RealmMyCourse) item).getUserId().contains(userId);
//        }else if(c == RealmMyLibrary.class ){
//            return ((RealmMyLibrary) item).getUserId().contains(userId);
//        }else if(c == RealmMyTeam.class){
//            return ((RealmMyTeam)item).getUserId().contains(userId);
//        }
//        return false;
//    }

    public static String[] getIds(Realm mRealm) {
        List<RealmMyLibrary> list = mRealm.where(RealmMyLibrary.class).findAll();
        String[] ids = new String[list.size()];
        int i = 0;
        for (RealmMyLibrary library : list
        ) {
            ids[i] = (library.getResource_id());
            i++;
        }
        return ids;
    }

    public static void removeDeletedResource(List<String> newIds, Realm mRealm) {
        String[] ids = getIds(mRealm);
        for (String id : ids
        ) {
            if (!newIds.contains(id)) {
                mRealm.where(RealmMyLibrary.class).equalTo("resourceId", id).findAll().deleteAllFromRealm();
            }
        }
    }

    public static JsonObject serialize(RealmMyLibrary personal, RealmUserModel user) {
        JsonObject object = new JsonObject();
        object.addProperty("title", personal.getTitle());
        object.addProperty("uploadDate", new Date().getTime());
        object.addProperty("createdDate", personal.getCreatedDate());
        object.addProperty("filename", FileUtils.getFileNameFromUrl(personal.getResourceLocalAddress()));
        object.addProperty("author", user.getName());
        object.addProperty("addedBy", user.getId());
        object.addProperty("medium", personal.getMedium());
        object.addProperty("description", personal.getDescription());
        object.addProperty("year", personal.getYear());
        object.addProperty("language", personal.getLanguage());
        object.add("subject", JsonUtils.getAsJsonArray(personal.getSubject()));
        object.add("level", JsonUtils.getAsJsonArray(personal.getLevel()));
        object.addProperty("resourceType", personal.getResourceType());
        object.addProperty("openWith", personal.getOpenWith());
        object.add("resourceFor", JsonUtils.getAsJsonArray(personal.getResourceFor()));
        object.addProperty("private", false);
        object.addProperty("isDownloadable", "");
        object.addProperty("sourcePlanet", user.getPlanetCode());
        object.addProperty("resideOn", user.getPlanetCode());
        object.addProperty("updatedDate", Calendar.getInstance().getTimeInMillis());
        object.addProperty("createdDate", personal.getCreatedDate());
        object.addProperty("androidId", NetworkUtils.getMacAddr());
        object.addProperty("deviceName", NetworkUtils.getDeviceName());
        object.addProperty("customDeviceName", NetworkUtils.getCustomDeviceName(MainApplication.context));
//        object.addProperty("mediaType", personal.getMediaType());
        return object;
    }


    public String getDownloaded() {
        return downloaded;
    }

    public void setDownloaded(String downloaded) {
        this.downloaded = downloaded;
    }

    public static void insertResources(JsonObject doc, Realm mRealm) {
        insertMyLibrary("", doc, mRealm);
    }


    public static void createStepResource(Realm mRealm, JsonObject res, String myCoursesID, String stepId) {
        insertMyLibrary("", stepId, myCoursesID, res, mRealm);

    }

    public static void insertMyLibrary(String userId, JsonObject doc, Realm mRealm) {
        insertMyLibrary(userId, "", "", doc, mRealm);
    }

    public String getMedium() {
        return medium;
    }

    public boolean isPrivate() {
        return isPrivate;
    }

    public void setPrivate(boolean aPrivate) {
        isPrivate = aPrivate;
    }

    public void setMedium(String medium) {
        this.medium = medium;
    }

    public static void createFromResource(RealmMyLibrary resource, Realm mRealm, String userId) {
        if (!mRealm.isInTransaction())
            mRealm.beginTransaction();
        resource.setUserId(userId);
        mRealm.commitTransaction();
    }

    public String get_id() {
        return _id;
    }

    public void set_id(String _id) {
        this._id = _id;
    }

    public static void insertMyLibrary(String userId, String stepId, String courseId, JsonObject doc, Realm mRealm) {

        String resourceId = JsonUtils.getString("_id", doc);
        SharedPreferences settings = MainApplication.context.getSharedPreferences(SyncActivity.PREFS_NAME, Context.MODE_PRIVATE);
        RealmMyLibrary resource = mRealm.where(RealmMyLibrary.class).equalTo("id", resourceId).findFirst();
        if (resource == null) {
            resource = mRealm.createObject(RealmMyLibrary.class, resourceId);
        }
//        if (!TextUtils.isEmpty(userId) ) {
        resource.setUserId(userId);
        resource.set_id(resourceId);
//        }
        if (!TextUtils.isEmpty(stepId)) {
            resource.setStepId(stepId);
        }
        if (!TextUtils.isEmpty(courseId)) {
            resource.setCourseId(courseId);
        }
        resource.set_rev(JsonUtils.getString("_rev", doc));

        resource.setResource_id(resourceId);
        resource.setTitle(JsonUtils.getString("title", doc));
        resource.setDescription(JsonUtils.getString("description", doc));
        if (doc.has("_attachments")) {
            JsonObject attachments = doc.get("_attachments").getAsJsonObject();
            JsonParser parser = new JsonParser();
            JsonElement element = parser.parse(String.valueOf(attachments));
            JsonObject obj = element.getAsJsonObject();
            Set<Map.Entry<String, JsonElement>> entries = obj.entrySet();
            for (Map.Entry<String, JsonElement> entry : entries) {
                if (entry.getKey().indexOf("/") < 0) {
                    resource.setResourceRemoteAddress(settings.getString("couchdbURL", "http://") + "/resources/" + resourceId + "/" + entry.getKey());
                    resource.setResourceLocalAddress(entry.getKey());
                    resource.setResourceOffline(FileUtils.checkFileExist(resource.getResourceRemoteAddress()));
                }
            }
        }
        resource.setFilename(JsonUtils.getString("filename", doc));
        resource.setAverageRating(JsonUtils.getString("averageRating", doc));
        resource.setUploadDate(JsonUtils.getString("uploadDate", doc));
        resource.setYear(JsonUtils.getString("year", doc));
        resource.setAddedBy(JsonUtils.getString("addedBy", doc));
        resource.setPublisher(JsonUtils.getString("publisher", doc));
        resource.setLinkToLicense(JsonUtils.getString("linkToLicense", doc));
        resource.setOpenWith(JsonUtils.getString("openWith", doc));
        resource.setArticleDate(JsonUtils.getString("articleDate", doc));
        resource.setKind(JsonUtils.getString("kind", doc));
        resource.setCreatedDate(JsonUtils.getLong("createdDate", doc));
        resource.setLanguage(JsonUtils.getString("language", doc));
        resource.setAuthor(JsonUtils.getString("author", doc));
        resource.setMediaType(JsonUtils.getString("mediaType", doc));
        resource.setResourceType(JsonUtils.getString("resourceType", doc));
        resource.setTimesRated(JsonUtils.getInt("timesRated", doc));
        resource.setMedium(JsonUtils.getString("medium", doc));
        resource.setResourceFor(JsonUtils.getJsonArray("resourceFor", doc), resource);
        resource.setSubject(JsonUtils.getJsonArray("subject", doc), resource);
        resource.setLevel(JsonUtils.getJsonArray("level", doc), resource);
        resource.setTag(JsonUtils.getJsonArray("tags", doc), resource);
        resource.setPrivate(JsonUtils.getBoolean("private", doc));
        resource.setLanguages(JsonUtils.getJsonArray("languages", doc), resource);
    }

    public String getResourceType() {
        return resourceType;
    }

    public void setResourceType(String resourceType) {
        this.resourceType = resourceType;
    }

    public long getCreatedDate() {
        return createdDate;
    }

    public void setCreatedDate(long createdDate) {
        this.createdDate = createdDate;
    }

    public JsonObject serializeResource() {
        JsonObject object = new JsonObject();
        object.addProperty("_id", _id);
        object.addProperty("_rev", _rev);
        object.addProperty("_rev", _rev);
        object.addProperty("need_optimization", need_optimization);

        object.add("resourceFor", getArray(resourceFor));
        object.addProperty("publisher", publisher);
        object.addProperty("linkToLicense", linkToLicense);
        object.addProperty("addedBy", addedBy);
        object.addProperty("uploadDate", uploadDate);
        object.addProperty("openWith", openWith);
        object.add("subject", getArray(subject));
        object.addProperty("kind", kind);
        object.addProperty("medium", medium);
        object.addProperty("language", language);
        object.addProperty("author", author);
        object.addProperty("sum", sum);
        object.addProperty("createdDate", uploadDate);
        object.add("level", getArray(level));
        object.add("languages", getArray(languages));
        object.add("tag", getArray(tag));
        object.addProperty("timesRated", timesRated);
        object.addProperty("year", year);
        object.addProperty("title", title);
        object.addProperty("averageRating", averageRating);
        object.addProperty("filename", filename);
        object.addProperty("mediaType", mediaType);
        object.addProperty("description", description);
        JsonObject ob = new JsonObject();
        ob.add(resourceLocalAddress, new JsonObject());
        object.add("_attachments", ob);
        return object;
    }

    public JsonArray getArray(RealmList<String> ar) {
        JsonArray sub = new JsonArray();
        if (ar != null) {
            for (String s : ar
            ) {
                sub.add(s);
            }
        }
        return sub;
    }

    public RealmList<String> getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        if (this.userId == null) {
            this.userId = new RealmList<>();
        }

        if (!this.userId.contains(userId) && !TextUtils.isEmpty(userId))
            this.userId.add(userId);
    }

    public String getResourceId() {
        return resourceId;
    }

    public void setResourceId(String resourceId) {
        this.resourceId = resourceId;
    }

    public String getResourceRemoteAddress() {
        return resourceRemoteAddress;
    }

    public void setResourceRemoteAddress(String resourceRemoteAddress) {
        this.resourceRemoteAddress = resourceRemoteAddress;
    }

    public String getResourceLocalAddress() {
        return resourceLocalAddress;
    }

    public void setResourceLocalAddress(String resourceLocalAddress) {
        this.resourceLocalAddress = resourceLocalAddress;
    }

    public Boolean getResourceOffline() {
        return resourceOffline;
    }

    public Boolean isResourceOffline() {
        return resourceOffline && TextUtils.equals(_rev, downloadedRev);
    }

    public void setResourceOffline(Boolean resourceOffline) {
        this.resourceOffline = resourceOffline;
    }

    public String getResource_id() {
        return resourceId;
    }

    public void setResource_id(String resource_id) {
        this.resourceId = resource_id;
    }

    public RealmList<String> getResourceFor() {
        return resourceFor;
    }


    public void setResourceFor(JsonArray array, RealmMyLibrary resource) {
        for (JsonElement s :
                array) {
            if (!(s instanceof JsonNull) && !resource.getResourceFor().contains(s.getAsString()))
                resource.getResourceFor().add(s.getAsString());
        }
    }

    public void setSubject(JsonArray array, RealmMyLibrary resource) {
        for (JsonElement s :
                array) {
            if (!(s instanceof JsonNull) && !resource.getSubject().contains(s.getAsString()))
                resource.getSubject().add(s.getAsString());
        }
    }

    public void setLevel(JsonArray array, RealmMyLibrary resource) {
        for (JsonElement s :
                array) {
            if (!(s instanceof JsonNull) && !resource.getLevel().contains(s.getAsString()))
                resource.getLevel().add(s.getAsString());
        }
    }

    public String getTranslationAudioPath() {
        return translationAudioPath;
    }

    public void setTranslationAudioPath(String translationAudioPath) {
        this.translationAudioPath = translationAudioPath;
    }

    public void setTag(JsonArray array, RealmMyLibrary resource) {
        for (JsonElement s :
                array) {
            if (!(s instanceof JsonNull) && !resource.getTag().contains(s.getAsString()))
                resource.getTag().add(s.getAsString());
        }
    }

    public void setLanguages(JsonArray array, RealmMyLibrary resource) {
        for (JsonElement s :
                array) {
            if (!(s instanceof JsonNull) && !resource.getLanguages().contains(s.getAsString()))
                resource.getLanguages().add(s.getAsString());
        }
    }


    public static CharSequence[] getListAsArray(RealmResults<RealmMyLibrary> db_myLibrary) {
        CharSequence[] array = new CharSequence[db_myLibrary.size()];
        for (int i = 0; i < db_myLibrary.size(); i++) {
            array[i] = db_myLibrary.get(i).getTitle();
        }
        return array;
    }

    public void setResourceFor(RealmList<String> resourceFor) {
        this.resourceFor = resourceFor;
    }

    public RealmList<String> getSubject() {
        return subject;
    }

    public void setUserId(RealmList<String> userId) {
        this.userId = userId;
    }

    public String getDownloadedRev() {
        return downloadedRev;
    }

    public void setDownloadedRev(String downloadedRev) {
        this.downloadedRev = downloadedRev;
    }

    public String getSubjectsAsString() {
        if ((subject).isEmpty()) {
            return "";
        }
        String str = "";
        for (String s : subject) {
            str += s + ", ";
        }
        return str.substring(0, str.length() - 1);
    }

    public void setSubject(RealmList<String> subject) {
        this.subject = subject;
    }

    public RealmList<String> getLevel() {
        return level;
    }

    public void setLevel(RealmList<String> level) {
        this.level = level;
    }

    public RealmList<String> getTag() {
        return tag;
    }
//
//    public String getTagAsString() {
//        StringBuilder s = new StringBuilder();
//        String[] tags = getTag().toArray(new String[getTag().size()]);
//        Arrays.sort(tags);
//        for (String tag : tags) {
//            s.append(tag).append(", ");
//        }
//        return s.toString();
//    }

    public static String listToString(RealmList<String> list) {
        StringBuilder s = new StringBuilder();
        for (String tag : list) {
            s.append(tag).append(", ");
        }
        return s.toString();

    }

    public String getCourseId() {
        return courseId;
    }

    public void setCourseId(String courseId) {
        this.courseId = courseId;
    }

    public String getStepId() {
        return stepId;
    }

    public void setStepId(String stepId) {
        this.stepId = stepId;
    }

    public void setTag(RealmList<String> tag) {
        this.tag = tag;
    }

    public RealmList<String> getLanguages() {
        return languages;
    }

    public void setLanguages(RealmList<String> languages) {
        this.languages = languages;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String get_rev() {
        return _rev;
    }

    public void set_rev(String _rev) {
        this._rev = _rev;
    }

    public boolean isNeed_optimization() {
        return need_optimization;
    }

    public void setNeed_optimization(boolean need_optimization) {
        this.need_optimization = need_optimization;
    }

    public String getPublisher() {
        return publisher;
    }

    public void setPublisher(String publisher) {
        publisher = publisher;
    }

    public String getLinkToLicense() {
        return linkToLicense;
    }

    public void setLinkToLicense(String linkToLicense) {
        this.linkToLicense = linkToLicense;
    }

    public String getAddedBy() {
        return addedBy;
    }

    public void setAddedBy(String addedBy) {
        this.addedBy = addedBy;
    }

    public String getUploadDate() {
        return uploadDate;
    }

    public void setUploadDate(String uploadDate) {
        this.uploadDate = uploadDate;
    }

    public String getOpenWith() {
        return openWith;
    }

    public void setOpenWith(String openWith) {
        this.openWith = openWith;
    }

    public String getArticleDate() {
        return articleDate;
    }

    public void setArticleDate(String articleDate) {
        this.articleDate = articleDate;
    }

    public String getKind() {
        return kind;
    }

    public void setKind(String kind) {
        this.kind = kind;
    }

    public String getLanguage() {
        return language;
    }

    public void setLanguage(String language) {
        this.language = language;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getYear() {
        return year;
    }

    public void setYear(String year) {
        this.year = year;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAverageRating() {
        return averageRating;
    }

    public void setAverageRating(String averageRating) {
        this.averageRating = averageRating;
    }

    public String getFilename() {
        return filename;
    }

    public void setFilename(String filename) {
        this.filename = filename;
    }

    public String getMediaType() {
        return mediaType;
    }

    public void setMediaType(String mediaType) {
        this.mediaType = mediaType;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getSendOnAccept() {
        return sendOnAccept;
    }

    public void setSendOnAccept(String sendOnAccept) {
        this.sendOnAccept = sendOnAccept;
    }

    public int getSum() {
        return sum;
    }

    public void setSum(int sum) {
        this.sum = sum;
    }

    public int getTimesRated() {
        return timesRated;
    }

    public void setTimesRated(int timesRated) {
        this.timesRated = timesRated;
    }

    public static List<String> save(JsonArray allDocs, Realm mRealm) {
        List<String> list = new ArrayList<>();
        for (int i = 0; i < allDocs.size(); i++) {
            JsonObject doc = allDocs.get(i).getAsJsonObject();
            doc = JsonUtils.getJsonObject("doc", doc);
            String id = JsonUtils.getString("_id", doc);
            if (!id.startsWith("_design")) {
                list.add(id);
                RealmMyLibrary.insertResources(doc, mRealm);
            }
        }
        return list;
    }

    public static JsonArray getMyLibIds(Realm realm, String userId) {
        List<RealmMyLibrary> myLibraries = realm.where(RealmMyLibrary.class).contains("userId", userId).findAll();
        JsonArray ids = new JsonArray();
        for (RealmMyLibrary lib : myLibraries
        ) {
            ids.add(lib.getId());
        }
        return ids;
    }

    @Override
    public String toString() {
        return title;
    }

    public void removeUserId(String id) {
        this.userId.remove(id);
    }

    public boolean needToUpdate() {
        return (getResourceLocalAddress() != null) && !getResourceOffline() || !(TextUtils.equals(get_rev(), getDownloadedRev()));
    }


//    public static Set<String> getLanguages(List<RealmMyLibrary> libraries) {
//        Set<String> list = new HashSet<>();
//        for (RealmMyLibrary li : libraries) {
//            if (!TextUtils.isEmpty(li.getLanguage()))
//                list.add(li.getLanguage());
//        }
//        return list;
//    }

    public static Set<String> getLevels(List<RealmMyLibrary> libraries) {
        Set<String> list = new HashSet<>();
        for (RealmMyLibrary li : libraries) {
            list.addAll(li.getLevel());
        }
        return list;
    }

    public static Set<String> getArrayList(List<RealmMyLibrary> libraries, String type) {
        Set<String> list = new HashSet<>();
        for (RealmMyLibrary li : libraries) {
            String s = type.equals("mediums") ? li.getMediaType() : li.getLanguage();
            if (!TextUtils.isEmpty(s))
                list.add(s);
        }
        return list;
    }

    public static Set<String> getSubjects(List<RealmMyLibrary> libraries) {
        Set<String> list = new HashSet<>();
        for (RealmMyLibrary li : libraries) {
            list.addAll(li.getSubject());
        }
        return list;
    }


}

File: app/src/main/java/org/ole/planet/myplanet/model/RealmMyLibrary.kt
package org.ole.planet.myplanet.model

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import androidx.room.PrimaryKey
import com.google.gson.JsonArray
import com.google.gson.JsonNull
import com.google.gson.JsonObject
import java.util.Calendar
import java.util.UUID
import org.ole.planet.myplanet.MainApplication.Companion.context
import org.ole.planet.myplanet.services.SharedPrefManager
import org.ole.planet.myplanet.utils.FileUtils
import org.ole.planet.myplanet.utils.JsonUtils
import org.ole.planet.myplanet.utils.NetworkUtils

/**
 * Room replacement for the former Realm `RealmMyLibrary` model (resources).
 *
 * The multi-valued primitive fields (`userId`, `resourceFor`, `subject`, `level`, `tag`,
 * `languages`, formerly `RealmList<String>`) become `List<String>?` stored as JSON via the shared
 * [org.ole.planet.myplanet.data.room.Converters]. `attachments` (formerly
 * `RealmList<RealmAttachment>`) — a value-object child never queried on its own — becomes an
 * embedded JSON `List<RealmAttachment>`. Shelf membership (`userId` list containment) is queried
 * with `LIKE` on the JSON column (see `MyLibraryDao`). The class name is kept so the wide resources
 * UI/repo surface is untouched; a later rename pass drops the `Realm` prefix. Persistence goes
 * through [org.ole.planet.myplanet.data.room.dao.MyLibraryDao].
 */
@Entity(tableName = "my_library", indices = [Index("_rev"), Index("titleNormal")])
open class RealmMyLibrary {
    @PrimaryKey
    @JvmField
    var id: String = ""
    @JvmField
    var _id: String? = null
    var userId: List<String>? = null
    var resourceRemoteAddress: String? = null
    var resourceLocalAddress: String? = null
    var resourceOffline: Boolean = false
    var resourceId: String? = null
    var _rev: String? = null
    var downloadedRev: String? = null
    var needsOptimization: Boolean = false
    var publisher: String? = null
    var linkToLicense: String? = null
    var addedBy: String? = null
    var uploadDate: String? = null
    var createdDate: Long = 0
    var openWith: String? = null
    var articleDate: String? = null
    var kind: String? = null
    var language: String? = null
    var author: String? = null
    var year: String? = null
    var medium: String? = null
    var title: String? = null
    var titleNormal: String? = null
    var averageRating: String? = null
    var filename: String? = null
    var mediaType: String? = null
    var resourceType: String? = null
    var description: String? = null
    var translationAudioPath: String? = null
    var sum: Int = 0
    var timesRated: Int = 0
    var resourceFor: List<String>? = null
    var subject: List<String>? = null
    var level: List<String>? = null
    var tag: List<String>? = null
    var languages: List<String>? = null
    var courseId: String? = null
    var stepId: String? = null
    var isPrivate: Boolean = false
    var privateFor: String? = null
    var attachments: List<RealmAttachment>? = null

    fun serializeResource(): JsonObject {
        return JsonObject().apply {
            addProperty("_id", _id)
            addProperty("_rev", _rev)
            addProperty("need_optimization", needsOptimization)
            add("resourceFor", resourceFor.toJsonArray())
            addProperty("publisher", publisher)
            addProperty("linkToLicense", linkToLicense)
            addProperty("addedBy", addedBy)
            addProperty("uploadDate", uploadDate)
            addProperty("openWith", openWith)
            add("subject", subject.toJsonArray())
            addProperty("kind", kind)
            addProperty("medium", medium)
            addProperty("language", language)
            addProperty("author", author)
            addProperty("sum", sum)
            addProperty("createdDate", uploadDate)
            add("level", level.toJsonArray())
            add("languages", languages.toJsonArray())
            add("tag", tag.toJsonArray())
            addProperty("timesRated", timesRated)
            addProperty("year", year)
            addProperty("title", title)
            addProperty("averageRating", averageRating)
            addProperty("filename", filename)
            addProperty("mediaType", mediaType)
            addProperty("description", description)
            val ob = JsonObject()
            resourceLocalAddress?.let { addr ->
                ob.add(addr, JsonObject())
            }
            add("_attachments", ob)
        }
    }

    private fun List<String>?.toJsonArray(): JsonArray {
        return JsonArray().apply {
            this@toJsonArray?.forEach { add(it) }
        }
    }

    fun setUserId(userId: String?) {
        if (userId.isNullOrBlank()) return
        val current = this.userId?.toMutableList() ?: mutableListOf()
        if (!current.contains(userId)) {
            current.add(userId)
        }
        this.userId = current
    }

    @Ignore
    fun isResourceOffline(): Boolean {
        return resourceOffline && _rev == downloadedRev
    }

    @get:Ignore
    val subjectsAsString: String
        get() = subject?.joinToString(", ") ?: ""

    override fun toString(): String {
        return title ?: ""
    }

    fun removeUserId(id: String?) {
        this.userId = this.userId?.filterNot { it == id }
    }

    fun needToUpdate(): Boolean {
        return !resourceOffline || resourceLocalAddress != null && _rev != downloadedRev
    }

    companion object {
        fun serialize(personal: RealmMyLibrary, user: RealmUser?): JsonObject {
            return JsonObject().apply {
                addProperty("title", personal.title)
                addProperty("uploadDate", System.currentTimeMillis())
                addProperty("createdDate", personal.createdDate)
                addProperty("filename", FileUtils.getFileNameFromUrl(personal.resourceLocalAddress))
                addProperty("author", personal.author ?: "")
                addProperty("addedBy", user?.id)
                addProperty("medium", personal.medium)
                addProperty("description", personal.description)
                addProperty("year", personal.year)
                addProperty("language", personal.language)
                addProperty("publisher", personal.publisher ?: "")
                addProperty("linkToLicense", personal.linkToLicense ?: "")
                add("subject", JsonUtils.getAsJsonArray(personal.subject))
                add("level", JsonUtils.getAsJsonArray(personal.level))
                addProperty("resourceType", personal.resourceType)
                addProperty("openWith", personal.openWith)
                addProperty("mediaType", personal.mediaType ?: "other")
                add("resourceFor", JsonUtils.getAsJsonArray(personal.resourceFor))
                addProperty("private", personal.isPrivate)
                if (personal.isPrivate && personal.privateFor != null) {
                    val privateForObj = JsonObject()
                    privateForObj.addProperty("teams", personal.privateFor)
                    add("privateFor", privateForObj)
                }
                addProperty("isDownloadable", true)
                addProperty("sourcePlanet", user?.planetCode)
                addProperty("resideOn", user?.planetCode)
                addProperty("updatedDate", Calendar.getInstance().timeInMillis)
                addProperty("createdDate", personal.createdDate)
                addProperty("androidId", NetworkUtils.getUniqueIdentifier())
                addProperty("deviceName", NetworkUtils.getDeviceName())
                addProperty("customDeviceName", NetworkUtils.getCustomDeviceName(context))
            }
        }

        data class InsertParams(
            val doc: JsonObject,
            val spm: SharedPrefManager,
            val userId: String? = "",
            val stepId: String? = "",
            val courseId: String? = "",
            val existing: RealmMyLibrary? = null
        )

        private fun JsonArray?.mergeInto(target: MutableList<String>) {
            this?.forEach { jsonElement ->
                val value = jsonElement.takeIf { it !is JsonNull }?.asString ?: return@forEach
                if (value !in target) {
                    target.add(value)
                }
            }
        }

        private fun mergedList(current: List<String>?, array: JsonArray?): List<String> {
            val target = current?.toMutableList() ?: mutableListOf()
            array.mergeInto(target)
            return target
        }

        /**
         * Builds/updates an unmanaged [RealmMyLibrary] from a CouchDB doc, merging into
         * [InsertParams.existing] when supplied (mirrors the former Realm find-or-create logic).
         */
        fun insertMyLibrary(params: InsertParams): RealmMyLibrary? {
            if (params.doc.entrySet().isEmpty()) return null
            val resourceId = JsonUtils.getString("_id", params.doc)
            val resource = params.existing ?: RealmMyLibrary().apply { id = resourceId }
            val wasPrivate = params.existing?.isPrivate == true
            val hadPrivateFor = params.existing?.privateFor
            val hadRev = params.existing?._rev
            val isLocalOnlyPrivate = hadRev.isNullOrBlank() && wasPrivate && !hadPrivateFor.isNullOrBlank()

            resource.apply {
                setUserId(params.userId)
                _id = resourceId
                if (!params.stepId.isNullOrBlank()) {
                    this.stepId = params.stepId
                }
                if (!params.courseId.isNullOrBlank()) {
                    this.courseId = params.courseId
                }
                _rev = JsonUtils.getString("_rev", params.doc)
                this.resourceId = resourceId
                val titleString = JsonUtils.getString("title", params.doc)
                title = titleString
                titleNormal = titleString.let {
                    java.text.Normalizer.normalize(it, java.text.Normalizer.Form.NFD)
                        .replace(Regex("\\p{InCombiningDiacriticalMarks}+"), "")
                        .lowercase(java.util.Locale.ROOT)
                }
                description = JsonUtils.getString("description", params.doc)
                if (params.doc.has("_attachments")) {
                    val attachmentsObj = params.doc["_attachments"].asJsonObject
                    val attachmentList = this.attachments?.toMutableList() ?: mutableListOf()

                    attachmentsObj.entrySet().forEach { (key, attachmentValue) ->
                        val attachmentObj = attachmentValue.asJsonObject
                        val realmAttachment = RealmAttachment().apply {
                            id = UUID.randomUUID().toString()
                            name = key
                            contentType = attachmentObj.get("content_type")?.asString
                            length = attachmentObj.get("length")?.asLong ?: 0
                            digest = attachmentObj.get("digest")?.asString
                            isStub = attachmentObj.get("stub")?.asBoolean == true
                            revpos = attachmentObj.get("revpos")?.asInt ?: 0
                        }
                        attachmentList.add(realmAttachment)

                        if (key.indexOf("/") < 0) {
                            resourceRemoteAddress = "${params.spm.getCouchdbUrl().ifEmpty { "http://" }}/resources/$resourceId/$key"
                            resourceLocalAddress = key
                            resourceOffline = FileUtils.checkFileExist(context, resourceRemoteAddress)
                            if (resourceOffline) {
                                downloadedRev = JsonUtils.getString("_rev", params.doc)
                            }
                        }
                    }
                    this.attachments = attachmentList
                }
                filename = JsonUtils.getString("filename", params.doc)
                averageRating = JsonUtils.getString("averageRating", params.doc)
                uploadDate = JsonUtils.getString("uploadDate", params.doc)
                year = JsonUtils.getString("year", params.doc)
                addedBy = JsonUtils.getString("addedBy", params.doc)
                publisher = JsonUtils.getString("publisher", params.doc)
                linkToLicense = JsonUtils.getString("linkToLicense", params.doc)
                openWith = JsonUtils.getString("openWith", params.doc)
                articleDate = JsonUtils.getString("articleDate", params.doc)
                kind = JsonUtils.getString("kind", params.doc)
                createdDate = JsonUtils.getLong("createdDate", params.doc)
                language = JsonUtils.getString("language", params.doc)
                author = JsonUtils.getString("author", params.doc)
                mediaType = JsonUtils.getString("mediaType", params.doc)
                resourceType = JsonUtils.getString("resourceType", params.doc)
                timesRated = JsonUtils.getInt("timesRated", params.doc)
                medium = JsonUtils.getString("medium", params.doc)
                resourceFor = mergedList(resourceFor, JsonUtils.getJsonArray("resourceFor", params.doc))
                subject = mergedList(subject, JsonUtils.getJsonArray("subject", params.doc))
                level = mergedList(level, JsonUtils.getJsonArray("level", params.doc))
                tag = mergedList(tag, JsonUtils.getJsonArray("tags", params.doc))
                if (!isLocalOnlyPrivate) {
                    isPrivate = JsonUtils.getBoolean("private", params.doc)
                    if (isPrivate && params.doc.has("privateFor")) {
                        val privateForElement = params.doc.get("privateFor")
                        if (privateForElement.isJsonObject) {
                            privateFor = privateForElement.asJsonObject.get("teams")?.asString
                        }
                    }
                }
                languages = mergedList(languages, JsonUtils.getJsonArray("languages", params.doc))
            }
            return resource
        }

        fun listToString(list: List<String>?): String {
            return list?.joinToString(", ") ?: ""
        }
    }
}

/**
 * Value-object attachment embedded (as JSON) in [RealmMyLibrary]. Never persisted or queried on
 * its own, so it is a plain class rather than a Room entity.
 */
open class RealmAttachment {
    var id: String? = null
    var name: String? = null
    var contentType: String? = null
    var length: Long = 0
    var digest: String? = null
    var isStub: Boolean = false
    var revpos: Int = 0
}
